[kombu] Add Kombu tracing

tebriel opened the pull request:
# Kombu Integration

Hi. We use kombu a lot, and are just starting to use tracing. I've started this pr (`tox -e '{py36}-kombu{42}'` passes) and I'm looking for feedback.
1. Would you accept this kind of PR once complete?
   1.  If not, how can I use this code outside of dd-trace-py?
2. Is this the right way to approach this?

Thanks!!

palazzem commented:
Hello @tebriel ! That's great! We definitely want to accept your PR inside our codebase and we'll proceed with the review to give you a couple of feedbacks.

I'll keep you updated once the PR is under review.
Thanks again!

tebriel commented:
Would really like some feedback on this @palazzem

palazzem commented:
Sure @tebriel ! it's in our radar and we plan to review and proceed with the merge very soon. Our current task (just for transparency) is rewriting all our documentation for Python. If any change is needed in this PR, we'll make the update so you don't need to spend time on Kombu documentation. I'll keep you updated, thanks!

Kyle-Verhoog commented on ddtrace/contrib/kombu/util.py:
```diff
@@ -0,0 +1,62 @@
+    name = 'Unknown'
+
+    # Definition of _publish has changed, we're in trouble if != 11
+    if len(args) == 11:
```
I don't think we need these checks. They don't guarantee us anything about the actual arguments. I think it suffices to ensure that the code is wrapped with a `try` and that we make sure to log any failures should they occur. 

This way if the version of `kombu` is not compatible with the integration then the user will be notified.

palazzem replied:
Yeah. In general if a version is not compatible it's fine to avoid tracing, but users must be notified with logs so they can report an issue and we can easily fix that.

Kyle-Verhoog commented on tests/contrib/kombu/test.py:
```diff
@@ -0,0 +1,105 @@
+    # ensure the old interface isn't broken, but doesn't trace
+    tracer = get_dummy_tracer()
+    TracedKombu = get_traced_kombu(tracer, "foo")
+    conn = TracedKombu("amqp://guest:guest@127.0.0.1:%s//" % (RABBITMQ_CONFIG['port'], ))
```
Let's stick to the library convention of using `"".format()` over string placeholders.

Kyle-Verhoog commented on tests/contrib/kombu/test.py:
```diff
@@ -0,0 +1,105 @@
+        conn.drain_events(timeout=2)
+
+    eq_(results[0], to_publish)
+    assert not tracer.writer.pop()
```
We should check these spans more thoroughly as done below in the `_assert_conn_traced` function.

Kyle-Verhoog commented:
Oh, to fix the CI, a job will have to be added for `kombu`. It would look very similar to the `redis` config in `.circleci/config.yml`:

https://github.com/DataDog/dd-trace-py/blob/02b096d3441540561b34a84a2af4ad0cc9cf710e/.circleci/config.yml#L699-L716

tebriel commented:
Thank you very much! I’ll look into this stuff this week. I appreciate the feedback.

On Aug 29, 2018, at 6:01 PM, Kyle Verhoog <notifications@github.com<mailto:notifications@github.com>> wrote:


Oh, to fix the CI, a job will have to be added for kombu. It would look very similar to the redis config in .circleci/config.yml:

https://github.com/DataDog/dd-trace-py/blob/02b096d3441540561b34a84a2af4ad0cc9cf710e/.circleci/config.yml#L699-L716

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/DataDog/dd-trace-py/pull/515#issuecomment-417146532>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAyJuGxn7KGioiJTPe6DaHnCvCNSJHxnks5uVytigaJpZM4VTjHw>.

palazzem commented on ddtrace/contrib/kombu/patch.py:
```diff
@@ -0,0 +1,91 @@
+    _w = wrapt.wrap_function_wrapper
+    _w('kombu', 'Producer._publish', traced_publish)
+    _w('kombu', 'Consumer.receive', traced_receive)
+    Pin(service="kombu", app="kombu", app_type="custom").onto(kombu.messaging.Producer)
```
for strings that can be configured by users, we should also add the Config API: https://github.com/DataDog/dd-trace-py/blob/fcfa5fd292d1dbfa941dca6c8f3f4c3fc85c7cfa/ddtrace/contrib/requests/patch.py#L15-L20

palazzem commented on ddtrace/contrib/kombu/patch.py:
```diff
@@ -0,0 +1,91 @@
+    with pin.tracer.trace('kombu.Consumer.receive', service=pin.service, span_type='kombu') as s:
+        # run the command
+        exchange = message.delivery_info['exchange']
+        s.resource = exchange
```
I'm not very familiar with Kombu, but what could be the cardinality of that? each single resource is computed as a metric, and if the cardinality is too high, the resource page results in something not really usable.

tebriel replied:
We have (what I consider) a very large redis cluster, pushing thousands of messages per second, and have about 250 exchanges. I'd be more worried if we were doing this with queues which can fan-out and bind to exchanges, but as it is, it should be a stable number, depending on the app. They don't typically have UUIDs or some other high cardinality item in their name.

Kyle-Verhoog replied:
According to the [documentation](http://docs.celeryproject.org/projects/kombu/en/latest/reference/kombu.html#exchange) it seems like exchanges should be appropriate to use for the resource.

palazzem commented on ddtrace/contrib/kombu/patch.py:
```diff
@@ -0,0 +1,91 @@
+    setattr(kombu, '_datadog_patch', True)
+
+    _w = wrapt.wrap_function_wrapper
+    _w('kombu', 'Producer._publish', traced_publish)
```
Patching an internal method may cause issues even with bugfix releases (i.e. method changes the name or the behavior). We must be sure our system doesn't crash users applications in case the internal method changes.

Also do we have other options to instrument that library instead of monkey-patching the main entrypoints?

tebriel replied:
I did this because it was quick/dirty/easier than mashing up the kwargs and args to find the desired fields, I'll clean this up, it was just to get *something* working.

palazzem commented on ddtrace/contrib/kombu/patch.py:
```diff
@@ -0,0 +1,91 @@
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    with pin.tracer.trace('kombu.Producer._publish', service=pin.service, span_type='kombu') as s:
```
The name of the operation couldn't be simply `kombu.publish` instead of `kombu.Producer._publish`? I think it's more readable from users stand point. What do you think?

palazzem commented on ddtrace/contrib/kombu/tracers.py:
```diff
@@ -0,0 +1,20 @@
+DEFAULT_SERVICE = 'kombu'
+
+
+@deprecated(message='Use patching instead (see the docs).', version='1.0.0')
```
Correct. The library is new so we just need to do the right thing, exposing (or not) all the methods users need.

palazzem commented on ddtrace/contrib/kombu/patch.py:
```diff
@@ -0,0 +1,91 @@
+from ...ext import kombu as kombux
+from ...utils.wrappers import unwrap
+from ...propagation.http import HTTPPropagator
+from .util import (
```
can we call this module `utils.py`? (plural)

palazzem commented on ddtrace/ext/kombu.py:
```diff
@@ -0,0 +1,10 @@
+TYPE = 'kombu'
+
+# net extension
+VHOST = 'out.vhost'
```
If you use `out.host`, our system will detect the tag and consider the execution as happened in another host (assuming the host is really different). There is a reason why you prefer `out.vhost`? Maybe it's part of the Kombu vocabulary?

tebriel replied:
A vhost is not a physical host, but a string that is part of the rabbit url. This is typically used to have multiple tenants in a rabbit cluster, say if you had two logically distinct apps that wanted to save on hardware costs. The deafult is `/` (urlencoded `%2F`), and this is typically the default.

tebriel commented on tox.ini:
```diff
@@ -70,6 +70,7 @@ envlist =
     {py34,py35,py36}-aiobotocore{02,03,04}
     {py34,py35,py36}-aiopg{012,013}
     {py27,py34,py35,py36}-redis{26,27,28,29,210}
+    {py36}-kombu{42}
```
Yes, I just picked 3.6 and 4.2 to get *something* working. I'll expand this.

Kyle-Verhoog commented:
@tebriel oh, you also need to actually list them for the jobs to be run:
here (with redis as an example again):
https://github.com/DataDog/dd-trace-py/blob/680433cec91051948d5e0336c6b82df0c2ef9b44/.circleci/config.yml#L860
and here:
https://github.com/DataDog/dd-trace-py/blob/680433cec91051948d5e0336c6b82df0c2ef9b44/.circleci/config.yml#L897

tebriel commented:
@palazzem I'm honestly not sure how they'll interact. I'll try to carve out some time tomorrow to read up on the Celery integration, and see if we can save some processing there or just flip one or the other. Otherwise I think I've addressed most of the other comments (hopefully I didn't miss any!)

tebriel commented:
@Kyle-Verhoog thanks very much for your quick help with the CircleCI Stuff. I think we're much closer to done at this point, if you have time to take a second look, I'd greatly appreciate it.

Kyle-Verhoog commented:
hey @tebriel, we're in the middle of releasing our OpenTracing then we'll circle back with you on this PR. Thanks for your patience and thorough follow-up!

tebriel commented on ddtrace/contrib/kombu/patch.py:
```diff
@@ -0,0 +1,114 @@
+    ).onto(kombu.messaging.Consumer)
+
+
+def unpatch():
```
I'm confused about this, because I copied what Redis Contrib does, and it does not do this. In fact celery uses a "patch_app/unpatch_app" while Redis just uses patch/unpatch. Can you help me understand the difference?

Kyle-Verhoog replied:
Yeah, sorry for the confusion. Our redis integration was one of the first added to the tracer and has a few issues, we're trying to correct these with our new integrations going forward 😄. 

Celery uses a `patch_app` and `unpatch_app` _in addition_ to `patch` and `unpatch` due to celery convention (I believe). Each integration must have `patch` and `unpatch` in order to be picked up by the patching mechanism.

Let me know if this all makes sense! 😄

tebriel replied:
Since the Kombu integration doesn't have an "app" convention, should I be doing something like that? Additionally, how do I unpin in the unpatch app? I'm not sure where to get the pin to remove it. Thanks for your help!

tebriel commented on ddtrace/contrib/kombu/patch.py:
```diff
@@ -0,0 +1,114 @@
+        return func(*args, **kwargs)
+
+
+def _get_tags(conn):
```
i thought that function was going to do more, and it never did. No reason for it to exist.

Kyle-Verhoog replied:
(this seems to be inherited from our redis integration which does the same thing, haha)

Kyle-Verhoog commented:
@tebriel have you gotten a chance to test this out in tandem with the celery integration? We can set up a demo app in https://github.com/DataDog/trace-examples and I'd be happy to assist with the testing.

tebriel commented on ddtrace/contrib/kombu/__init__.py:
```diff
@@ -0,0 +1,34 @@
+"""Instrument kombu to report AMQP messaging.
```
Added this.

tebriel commented on ddtrace/contrib/kombu/__init__.py:
```diff
@@ -0,0 +1,34 @@
+    conn.connect()
+    task_queue = kombu.Queue('tasks', kombu.Exchange('tasks'), routing_key='tasks')
+    to_publish = {'hello': 'world'}
+    producer.publish(to_publish,
```
Fixed that, good find!

tebriel commented:
python on my system is in a really bad state, i'm going to have to rely on circle-ci to test until I can get that sorted out

brettlangdon commented:
@tebriel if you have `docker-compose` available on your machine, we have a Docker container we all use locally for testing (the same one we use in CircleCI).

https://github.com/DataDog/dd-trace-py#running-tests-in-docker

If you need any help or run into any trouble using that container, please let us know!

tebriel commented:
> @tebriel if you have `docker-compose` available on your machine, we have a Docker container we all use locally for testing (the same one we use in CircleCI).
> 
> https://github.com/DataDog/dd-trace-py#running-tests-in-docker
> 
> If you need any help or run into any trouble using that container, please let us know!

Actually it was more helpful to use the circleci cli (glad you had instructions for that too!). My tests were passing, but my circleci config was no longer valid (major changes to this repo since I started this PR). Thanks! Passes locally, waiting for CircleCI to agree.

tebriel commented:
@Kyle-Verhoog @palazzem So, my tests are passing, but Cassandra and OpenTracing failed. It looks like I don't have the ability to re-kick off CircleCI. Sorry for the long response time on addressing PR comments, but I think (hope!) I addressed this round. Let me know what else I can do to improve, and thanks for your time!

brettlangdon commented:
@tebriel I was just able to manually run the failed jobs and all good

tebriel commented:
I’ll get it done in the next few hours! Thank you very much!

On Oct 31, 2018, at 7:06 AM, Kyle Verhoog <notifications@github.com<mailto:notifications@github.com>> wrote:


@Kyle-Verhoog commented on this pull request.

Hey @tebriel<https://github.com/tebriel> thanks for your perseverance 😄.

Just some quote nits. I re-ran with the demo app here: DataDog/trace-examples#39<https://github.com/DataDog/trace-examples/pull/39> and all looks good to me! Since you are running a fairly large kombu instance yourselves, if the traces are insightful for you then it's good to go on my end!

Since it's just these small changes I'm going to set the base of the branch to 0.16-dev to get it in with our next release.

Once everything's updated then I'll give it the ✅!

Thanks!

________________________________

In ddtrace/contrib/kombu/patch.py<https://github.com/DataDog/dd-trace-py/pull/515#discussion_r229646092>:

> +    _w = wrapt.wrap_function_wrapper
+    # We wrap the _publish method because the publish method:
+    # *  defines defaults in its kwargs
+    # *  potentially overrides kwargs with values from self
+    # *  extracts/normalizes things like exchange
+    _w(kombux.TYPE, 'Producer._publish', traced_publish)
+    _w(kombux.TYPE, 'Consumer.receive', traced_receive)
+    Pin(
+        service=config.kombu['service_name'],
+        app="kombu",
+        app_type=AppTypes.worker,
+    ).onto(kombu.messaging.Producer)
+
+    Pin(
+        service=config.kombu['service_name'],
+        app="kombu",


⬇️ Suggested change

-        app="kombu",
+        app='kombu',


________________________________

In ddtrace/contrib/kombu/patch.py<https://github.com/DataDog/dd-trace-py/pull/515#discussion_r229646142>:

> +    of Kombu. However, it means that any "import kombu.Connection" won't be instrumented.
+    """
+    if getattr(kombu, '_datadog_patch', False):
+        return
+    setattr(kombu, '_datadog_patch', True)
+
+    _w = wrapt.wrap_function_wrapper
+    # We wrap the _publish method because the publish method:
+    # *  defines defaults in its kwargs
+    # *  potentially overrides kwargs with values from self
+    # *  extracts/normalizes things like exchange
+    _w(kombux.TYPE, 'Producer._publish', traced_publish)
+    _w(kombux.TYPE, 'Consumer.receive', traced_receive)
+    Pin(
+        service=config.kombu['service_name'],
+        app="kombu",


⬇️ Suggested change

-        app="kombu",
+        app='kombu',


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/DataDog/dd-trace-py/pull/515#pullrequestreview-170191279>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAyJuOtu9tneL3BkJsRdSLTG8M_C8rE1ks5uqYRPgaJpZM4VTjHw>.

tebriel commented:
@Kyle-Verhoog sorry it took me a while to get back to it. Squashed and Rebased off of `0-16.dev`

tebriel commented:
Integration test failed, but all the rest passed. Can someone re-run it? Is it known to be unstable?

Kyle-Verhoog commented:
@tebriel yep, it's unstable. We'll re-run it for you 😄.
